mavros/src/plugins/local_position.cpp

```diff
@@ -63,36 +69,60 @@ class LocalPositionPlugin : public MavRosPlugin {
 
 	ros::Publisher local_position;
 	ros::Publisher local_velocity;
+	ros::Publisher local_odom;
 
 	std::string frame_id;		//!< frame for Pose
 	std::string tf_frame_id;	//!< origin for TF
 	std::string tf_child_frame_id;	//!< frame for TF
+	std::string tf_base_link_frame_id; //!< frame for TF (specifically Odometry message)
 	bool tf_send;
 
 	void handle_local_position_ned(const mavlink_message_t *msg, uint8_t sysid, uint8_t compid) {
 		mavlink_local_position_ned_t pos_ned;
 		mavlink_msg_local_position_ned_decode(msg, &pos_ned);
 
-		auto position = UAS::transform_frame_ned_enu(Eigen::Vector3d(pos_ned.x, pos_ned.y, pos_ned.z));
-		auto velocity = UAS::transform_frame_ned_enu(Eigen::Vector3d(pos_ned.vx, pos_ned.vy, pos_ned.vz));
+		auto enu_position = UAS::transform_frame_ned_enu(Eigen::Vector3d(pos_ned.x, pos_ned.y, pos_ned.z));
+		auto enu_velocity = UAS::transform_frame_ned_enu(Eigen::Vector3d(pos_ned.vx, pos_ned.vy, pos_ned.vz));
+
 		auto orientation = uas->get_attitude_orientation();
 		auto angular_velocity = uas->get_attitude_angular_velocity();
 
+		// Here the orientation quaternion is describing the rotation from the body frame 
+		// to the ENU frame.  We also have the velocity expressed in the ENU frame.  We
+		// Therefore need to calculate the inverse quaternion to find the transformation
+		// from the ENU frame to the body frame.  Then we can apply this rotation to the 
+		// velocity expressed in the ENU frame to get the body frame velocity.  
+		auto enu_to_body_quad = Eigen::Quaterniond(orientation.w, orientation.x, orientation.y, orientation.z).inverse();
+		auto enu_angular_velocity = Eigen::Vector3d(angular_velocity.x, angular_velocity.y, angular_velocity.z);
+		Eigen::Transform<double, 3, Eigen::Affine> enu_to_body_rot(enu_to_body_quad);
```

Comment:
I think `Eigen::Affine3d` is the same type.
Also you may use `eigen_conversions/eigen_msg.h` for `enu_*`.